eventlog/eventlog.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2020 GRAIL, Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+// Package eventlog provides logging of semi-structured events, particularly in
+// service of downstream analysis, e.g. when machines are started, when a user
+// issues a command, when failures happen.
+//
+// For example, you can log events to CloudWatch Logs:
+//
+//  sess := session.NewSession()
+//  cw := cloudwatchlogs.New(sess)
+//  e := NewCloudWatchEventer(cw, "myLogGroup", "myLogStream")
+//  e.Event("rpcRetry", "org", "GRAIL", "retry", 0, "maxRetry", 10)
+//  e.Event("machineStopped", "addr", "192.168.1.1", "duration", 3600.0, "startTime": 1584140534)
+//
+// These events can now be analyzed and monitored using CloudWatch Logs tooling.
+package eventlog
+
+import (
+	"github.com/grailbio/base/config"
+)
+
+func init() {
+	config.Register("eventer/nop", func(constr *config.Constructor) {
+		constr.Doc = "eventer/nop configures a no-op event logger"
+		constr.New = func() (interface{}, error) {
+			return Nop{}, nil
+		}
+	})
+	// We do the most conservative thing by default, making event logging a
+	// no-op.
+	config.Default("eventer", "eventer/nop")
+}
+
+// Eventer is called to log events.
+type Eventer interface {
```

Comment:
I thought I might see some sort of Eventer mock that implements this interface and can be used during tests that just stores the logged events locally for inspection during testing.

Reply:
Yes; right now I think of this more like (best-effort) logging, just formatted for easy parsing and with data we're specifically looking to collect. I don't expect people to test the behavior just like they don't test that their programs write logs properly, even if there are scripts out there that do some ad hoc parsing of logs. If it evolves into something on which rely more critically, we'll definitely have to revisit.